File: Road_C/Src_PYTHON/macfind_c/bitmask.py
"""Cling-style 64-bit character bitmask for O(1) candidate pre-filtering.

Each set bit records that a given character *class* appears somewhere in the
(lowercased) string. Bit layout is shared with the other language
implementations in this repo (see ``open-source-analysis.md`` §3.3 and the Go
peer at ``Road_C/Src_GO/internal/bitmask/bitmask.go``), so indexes stay
conceptually comparable across languages:

    Bits 0-25:  letters a-z
    Bits 26-35: digits 0-9
    Bit 36:     '.'
    Bit 37:     '-'
    Bit 38:     '_'

A query can only match an entry if every bit of the query mask is also set in
the entry mask, i.e. ``entry & query == query``. One ``AND`` discards the vast
majority of non-matching entries.
"""

from __future__ import annotations

import numpy as np

BIT_DOT = 36
BIT_DASH = 37
BIT_UNDERSCORE = 38

# Precomputed byte -> mask table (256 entries). Index by the *lowercased* byte.
_BYTE_MASK = np.zeros(256, dtype=np.uint64)
for _c in range(ord("a"), ord("z") + 1):
    _BYTE_MASK[_c] = np.uint64(1) << np.uint64(_c - ord("a"))
for _c in range(ord("0"), ord("9") + 1):
    _BYTE_MASK[_c] = np.uint64(1) << np.uint64(26 + (_c - ord("0")))
_BYTE_MASK[ord(".")] = np.uint64(1) << np.uint64(BIT_DOT)
_BYTE_MASK[ord("-")] = np.uint64(1) << np.uint64(BIT_DASH)
_BYTE_MASK[ord("_")] = np.uint64(1) << np.uint64(BIT_UNDERSCORE)

# Fold uppercase ASCII onto lowercase so callers with mixed-case input still map
# to the same class bit.
for _c in range(ord("A"), ord("Z") + 1):
    _BYTE_MASK[_c] = _BYTE_MASK[_c + (ord("a") - ord("A"))]

del _c


def of_bytes(data: bytes) -> np.uint64:
    """Return the character-class mask for ``data`` (a lowercased path/query)."""
    if not data:
        return np.uint64(0)
    arr = np.frombuffer(data, dtype=np.uint8)
    # Bitwise-OR of the per-byte masks. ``np.bitwise_or.reduce`` over the lookup.
    return np.bitwise_or.reduce(_BYTE_MASK[arr])


def of_str(s: str) -> np.uint64:
    """Return the character-class mask for ``s`` (lowercased on the fly)."""
    return of_bytes(s.lower().encode("utf-8", "ignore"))


def matches(entry: int, query: int) -> bool:
    """True if an entry could contain every character class the query needs."""
    return (int(entry) & int(query)) == int(query)

File: Road_C/Src_GO/internal/bitmask/bitmask.go
// Package bitmask implements the Cling-style 64-bit character bitmask used for
// O(1) candidate pre-filtering. Each set bit records that a given character
// class appears somewhere in the (lowercased) string.
//
// Bit layout (see open-source-analysis.md §3.3):
//
//	Bits 0-25:  letters a-z
//	Bits 26-35: digits 0-9
//	Bit 36:     '.'
//	Bit 37:     '-'
//	Bit 38:     '_'
//
// A query can only match an entry if every bit of the query's mask is also set
// in the entry's mask, i.e. `entryMask & queryMask == queryMask`. This lets us
// discard the vast majority of non-matching entries with a single AND.
package bitmask

const (
	bitDot        = 36
	bitDash       = 37
	bitUnderscore = 38
)

// Of computes the character-class bitmask for s. The input is assumed to be
// already lowercased (the index stores lowercased paths); any byte outside the
// tracked classes is ignored.
func Of(s string) uint64 {
	var m uint64
	for i := 0; i < len(s); i++ {
		m |= bitOf(s[i])
	}
	return m
}

// OfLower lowercases ASCII letters on the fly, for callers that hold mixed-case
// input (e.g. a raw user query).
func OfLower(s string) uint64 {
	var m uint64
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 'A' && c <= 'Z' {
			c += 'a' - 'A'
		}
		m |= bitOf(c)
	}
	return m
}

func bitOf(c byte) uint64 {
	switch {
	case c >= 'a' && c <= 'z':
		return 1 << (c - 'a')
	case c >= '0' && c <= '9':
		return 1 << (26 + (c - '0'))
	case c == '.':
		return 1 << bitDot
	case c == '-':
		return 1 << bitDash
	case c == '_':
		return 1 << bitUnderscore
	}
	return 0
}

// Matches reports whether an entry with mask entry could contain every
// character class required by query.
func Matches(entry, query uint64) bool {
	return entry&query == query
}
